core/src/ldml/ldml_processor.cpp

```diff
@@ -59,23 +59,130 @@ namespace {
   };
 }
 
+/**
+ * @brief Usage: `KMXPLUS_PRINTF(("str: %s\n", "something"));`
+ * Note double parens
+ * \def KMXPLUS_DEBUG
+ */
+#if KMXPLUS_DEBUG
+#include <iostream>
+#define KMXPLUS_NOTEQUAL(expect, actual) std::cerr << __FILE__ << ":" << __LINE__ << ": ASSERT FAILED: " \
+              << #actual << "=" << (actual) << ", expected " << (expect) << std::endl
+#define KMXPLUS_PRINTLN(msg) std::cerr  << __FILE__ << ":" << __LINE__ << ": " msg << std::endl;
+#else
+#define KMXPLUS_NOTEQUAL(expect, actual)
+#define KMXPLUS_PRINTLN(msg)
+#endif
+
 namespace km {
 namespace kbp {
 
 ldml_processor::ldml_processor(path const & kb_path, const std::vector<uint8_t> &data)
 : abstract_processor(
     keyboard_attributes(kb_path.stem(), KM_KBP_LMDL_PROCESSOR_VERSION, kb_path.parent(), {})
-  ), _valid(false), rawdata(data) // TODO-LDML: parse the data, don't just copy it
+  ), _valid(false), vkey_to_string()
 {
+
+/**
+ * Assert something, just in this function
+ * If fails, print a message and exit as invalid
+ * \def KMXPLUS_ASSERT
+ */
+#define KMXPLUS_ASSERT(expect,actual) if ((expect) != (actual)) { \
+    KMXPLUS_NOTEQUAL(expect, actual); \
+    _valid = false; \
+    return; \
+  }
+  const kmx::COMP_KMXPLUS_SECT *sect = nullptr;
+  const kmx::COMP_KMXPLUS_STRS *strs = nullptr;
+  const kmx::COMP_KMXPLUS_KEYS *keys = nullptr;
+
   // TODO-LDML: load the file from the buffer (KMXPlus format)
   // Note: kb_path is essentially debug metadata here
-  assert(data.size() != 0);
+  KMXPLUS_ASSERT(true, data.size() > sizeof(kmx::COMP_KEYBOARD_EX));
 
+  // Locate the structs here, but still retain ptrs to the raw structs.
   const kmx::PCOMP_KEYBOARD comp_keyboard = (kmx::PCOMP_KEYBOARD)data.data();
```

Comment:
```suggestion
  const kmx::COMP_KEYBOARD* comp_keyboard = (kmx::COMP_KEYBOARD*)data.data();
```

Can we use `reinterpret_cast` here also? Suggest using pointer type styling consistently (`COMP_FOO*` everywhere or `PCOMP_FOO` everywhere).